lib/Dialect/QUIR/Transforms/BreakReset.cpp

```diff
@@ -131,13 +143,149 @@ void BreakResetPass::runOnOperation() {
+  // putCallGatesAndMeasuresIntoCircuit option is true
+  mlir::ModuleOp moduleOp = getOperation();
+  uint circNum = 0;
+  while (!measureList.empty()) {
+    MeasureOp measOp = dyn_cast<MeasureOp>(measureList.front());
+    putMeasureInCircuit(moduleOp, measOp, circNum);
+    circNum++;
+    measureList.pop_front();
+  }
+  while (!callGateList.empty()) {
+    CallGateOp callGateOp = dyn_cast<CallGateOp>(callGateList.front());
+    putCallGateInCircuit(moduleOp, callGateOp, circNum);
+    circNum++;
+    callGateList.pop_front();
+  }
 } // BreakResetPass::runOnOperation
 
+void BreakResetPass::putCallGateInCircuit(ModuleOp moduleOp,
+                                          mlir::quir::CallGateOp callGateOp,
+                                          uint circNum) {
+  mlir::func::FuncOp mainFunc =
+      dyn_cast<mlir::func::FuncOp>(quir::getMainFunction(moduleOp));
+  assert(mainFunc && "could not find the main func");
+  mlir::OpBuilder builder(mainFunc);
+  mlir::Location location = mainFunc.getLoc();
+
+  auto circOp = builder.create<CircuitOp>(
+      mainFunc.getLoc(), "reset_x_" + std::to_string(circNum++),
+      builder.getFunctionType(
+          /*inputs=*/ArrayRef<Type>(),
+          /*results=*/ArrayRef<Type>()));
+
+  circOp.addEntryBlock();
+  OpBuilder circuitBuilder = OpBuilder::atBlockBegin(&circOp.getBody().front());
+
+  uint argumentIndex = 0;
+  llvm::SmallVector<Type> inputTypes;
+  llvm::SmallVector<Value> inputValues;
+  llvm::SmallVector<Type> outputTypes;
+  llvm::SmallVector<Value> outputValues;
+
+  for (auto operand : callGateOp->getOperands()) {
+    inputTypes.push_back(operand.getType());
+    inputValues.push_back(operand);
+    circOp.insertArgument(argumentIndex, operand.getType(), {},
+                          operand.getLoc());
+    argumentIndex++;
+  }
+
+  auto newCallGateOp = circuitBuilder.create<CallGateOp>(
+      location, StringRef("x"), TypeRange{}, circOp.getArguments());
+
+  outputValues.append(newCallGateOp->result_begin(),
+                      newCallGateOp->result_end());
+  outputTypes.append(newCallGateOp->result_type_begin(),
+                     newCallGateOp->result_type_end());
+
+  circOp.setType(circuitBuilder.getFunctionType(
+      /*inputs=*/ArrayRef<Type>(inputTypes),
+      /*results=*/ArrayRef<Type>(outputTypes)));
+
+  auto returnOp =
+      circuitBuilder.create<mlir::quir::ReturnOp>(location, ValueRange({}));
+  returnOp->insertOperands(0, ValueRange(outputValues));
+
+  auto callCircOp = builder.create<mlir::quir::CallCircuitOp>(
+      circOp->getLoc(), circOp.getName(), TypeRange(outputTypes),
+      ValueRange(inputValues));
+
+  callCircOp->moveBefore(callGateOp);
+  callGateOp->erase();
+}
+
+void BreakResetPass::putMeasureInCircuit(ModuleOp moduleOp,
+                                         mlir::quir::MeasureOp measureOp,
+                                         uint circNum) {
+  mlir::func::FuncOp mainFunc =
```

Comment:
I think you could factor out the circuit / call_circuit creation boilerplate into a single common function.